### PR TITLE
Replace z3 jar files with KSMT dependency

### PR DESCRIPTION
**Pull request by CaelmBleidd**
## Description

I've replaced prebuilt Z3 jars with KSMT dependency. It has three consequences: 
1. We no longer have to hold `jar` along with our source code in our repository and build it by hand
2. We have an updated version of Z3 since KSMT `0.4.3` works with `Z3 4.11.2`. Because of that, we have to generalize Z3 expressions with types
3. Now we have support for Apple Silicon processors

### Automated tests

Since it is a global change, all automatic tests are affected by it.

### Manual tests

It was tested with automatic tests on the pipeline on Ubuntu, locally on Windows and Mac with both M1 Pro and Intel-based processors.

## Self-check list

- [x] I've set the proper **labels** for my PR (at least, for category and component).
- [x] PR **title** and **description** are clear and intelligible.
- [ ] I've added enough **comments** to my code, particularly in hard-to-understand areas.
- [x] The functionality I've repaired, changed or added is covered with **automated tests**.
- [ ] **Manual tests** have been provided optionally.
- [ ] The **documentation** for the functionality I've been working on is up-to-date.

### CODE REVIEW

File: utbot-framework-test/src/test/kotlin/org/utbot/framework/z3/extension/Z3ExtensionForTests.kt [hunk 1]
```diff
@@ -27,20 +27,24 @@ infix fun BoolExpr.implies(other: BoolExpr): BoolExpr = context.mkImplies(this,
...
+infix fun ArithExpr<*>.gt(other: Int): BoolExpr = context.mkGt(this, context.mkInt(other))
 
-infix fun ArithExpr.`=`(other: Int): BoolExpr = context.mkEq(this, context.mkInt(other))
+infix fun ArithExpr<*>.`=`(other: Int): BoolExpr = context.mkEq(this, context.mkInt(other))
 
-operator fun ArrayExpr.get(index: IntExpr): Expr = context.mkSelect(this, index)
-operator fun ArrayExpr.get(index: Int): Expr = this[context.mkInt(index)]
-fun ArrayExpr.set(index: IntExpr, value: Expr): ArrayExpr = context.mkStore(this, index, value)
-fun ArrayExpr.set(index: Int, value: Expr): ArrayExpr = set(context.mkInt(index), value)
+operator fun  ArrayExpr<*, *>.get(index: IntExpr): Expr<*> = context.mkSelect(this, index.cast())
```
Comment (Damtev) [inline utbot-framework-test/src/test/kotlin/org/utbot/framework/z3/extension/Z3ExtensionForTests.kt:35]: Some redundant whitespaces her and below

File: utbot-framework-test/src/test/kotlin/org/utbot/framework/z3/extension/Z3ExtensionForTests.kt [hunk 2]
```diff
@@ -27,20 +27,24 @@ infix fun BoolExpr.implies(other: BoolExpr): BoolExpr = context.mkImplies(this,
...
+fun  ArrayExpr<*, *>.set(index: Int, value: Expr<*>):  ArrayExpr<*, *> = set(context.mkInt(index), value)
 
-operator fun SeqExpr.plus(other: SeqExpr): SeqExpr = context.mkConcat(this, other)
-operator fun SeqExpr.plus(other: String): SeqExpr = context.mkConcat(context.mkString(other))
+operator fun SeqExpr<*>.plus(other: SeqExpr<*>): SeqExpr<*> = context.mkConcat(this, other.cast())
+operator fun SeqExpr<*>.plus(other: String): SeqExpr<*> = context.mkConcat(context.mkString(other))
 
-infix fun SeqExpr.`=`(other: String): BoolExpr = context.mkEq(this, context.mkString(other))
+@Suppress("UNCHECKED_CAST")
+fun <T : Sort, R : Sort> Expr<T>.cast(): Expr<R> = this as Expr<R>
```
Comment (Damtev) [inline utbot-framework-test/src/test/kotlin/org/utbot/framework/z3/extension/Z3ExtensionForTests.kt:44]: You introduced the same function in `utbot-framework/src/main/kotlin/org/utbot/engine/pc/Util.kt`. Can we use it here?

File: utbot-testing/build.gradle [hunk 3]
```diff
@@ -48,4 +47,7 @@ dependencies {
 
     testImplementation group: 'org.mockito', name: 'mockito-inline', version: mockitoInlineVersion
     testImplementation group: 'org.apache.logging.log4j', name: 'log4j-core', version: log4j2Version
+
+    implementation group: 'com.github.UnitTestBot.ksmt', name: 'ksmt-core', version: ksmtVersion
```
Comment (Damtev) [inline utbot-testing/build.gradle:51]: Do we really need the `core` dependency here?

**Review by Damtev — APPROVED** [review]
LGTM

File: utbot-framework-test/src/test/kotlin/org/utbot/framework/z3/extension/Z3ExtensionForTests.kt (hunk 2 above)
Comment (CaelmBleidd) [reply utbot-framework-test/src/test/kotlin/org/utbot/framework/z3/extension/Z3ExtensionForTests.kt:44]: No, it is a different module

**Pull request merged**